rawfile/utils/snapshot_manager.py

```diff
@@ -0,0 +1,115 @@
+from pathlib import Path
+import time
+from typing import List
+from dataclasses import dataclass
+import consts
+from utils.commands import run
+
+from utils.devices import device_to_mountpoint
+from utils.lock import VolLock
+from utils.rawfile import attached_loops, img_file, snapshots_dir
+from glob import glob
+
+
+@dataclass
+class Snapshot:
+    name: str
+    volume_id: str
+    size_bytes: int
+    creation_time: float
+    ready: bool = True
+
+
+class SnapshotManager:
+    def _get_snapshot_path(self, volume_id: str, name: str) -> Path:
+        return snapshots_dir(volume_id) / f"{name}.img"
+
+    def create_snapshot(
+        self,
+        volume_id: str,
+        name: str,
+        copy_on_write: bool = False,
+        freeze_fs: bool = False,
+    ) -> Snapshot:
+        with VolLock(volume_id):
+            file = img_file(volume_id)
+            loop_devs = attached_loops(file.as_posix())
+            snap_path = self._get_snapshot_path(volume_id, name)
+            snap_path.parent.mkdir(parents=True, exist_ok=True)
+            Path(f"{snap_path}.creating").touch()
+            Path(snap_path).unlink(missing_ok=True)
+            mountpoints = []
+            snap_inuse = freeze_fs or copy_on_write
+            if (loop_devs and len(loop_devs)) and (not snap_inuse):
+                raise  # FIXME: Raise a better Exception
+            if loop_devs and len(loop_devs):
+                for dev in loop_devs:
+                    mp = device_to_mountpoint(dev, first_only=False)
+                    if mp:
+                        mountpoints.extend(mp)
+            try:
+                if freeze_fs and mountpoints:
+                    for mp in mountpoints:
+                        run(f"fsfreeze --freeze {mp}", check=True)
+
+                reflink = "always" if copy_on_write else "never"
+                cmd = f"cp --sparse=always --reflink={reflink} {file} {snap_path}"
+                run(cmd, check=True)
+
+                creation_time = time.time()
+                Path(f"{snap_path}.creating").unlink(missing_ok=True)
+                return Snapshot(
+                    name=name,
+                    volume_id=volume_id,
+                    size_bytes=snap_path.stat().st_size,
+                    creation_time=creation_time,
+                )
+            except Exception as e:
+                Path(snap_path).unlink(missing_ok=True)
+                raise e
+            finally:
+                if freeze_fs and mountpoints:
+                    for mp in mountpoints:
+                        run(f"fsfreeze --unfreeze {mp}", check=True)
```

Comment:
in case of a crash who unfreezes, is it the task manager by retrying the create again?
and if so, what happens if snapshot is deleted meanwhile?

Reply:
We have graceful shutdown, but we could add an static job on startup to unfreeze everything

Reply:
And also what happens if we receive another csi call whilst we have a copy manager?
Ex: 
- snapshot may be taken whilst volume is unpublished, then we copy... and then we receive a controllerPublish
- snapshot is taken whilst volume is staged on a node, then we receive a nodePublish or nodeUnpublish
- etc

Reply:
We are using volume locks in all of them So other Operations are blocked when doing snapshot

Reply:
What if the pod restarts?

Reply:
a static unfreeze on startup should fix that, that's the only idea that I have

Reply:
Why doing FSFreeze when we can Set entire blockdev to be RO?
It is easier to manage and also `attach_loop` will set them to RW automatically

Reply:
@tiagolobocastro the only this that is left is this one :)

Reply:
I have handled crashes during snapshot creation

I think it is better to handle all of the stuck tasks, So we will have safe clones later on (I will refactor) :)